Feature/report generation extenson

Elch2142 opened the pull request:
- Extends the report generation by the following features:
- Sorting options for reports: Each report can be sorted by either task, user, team or report week
- Added option for author role users to generate reports containing their own progresses
- Chrome browser based Flutter debugging has been reenabled by altering cookie settings for development builds debugging

Commit:
Updated scripts to use regular flutter channel

- Web functionality is no longer only in the beta channel
- Changed flutter path to default location when using the snap image

Commit:
Added new classes for upcoming report functionality

- Added Report and ReportSection as container classes containing the report data
- Each Report instance contains a progress stream of all progresses which will be listed in the report
- Each ReportSection instance contains a subset of theses Progresses together with a section title
- The sections contained within a Report instance are created by sorting the Progress stream
- Sorting will be done by the methods presented in the ReportSorter interface

Commit:
Restructured ReportSorter methods, added ReportSorter unit test

- Changed ReportSorter methods to work on a Collection instead of a Stream
- This is necessary because streams are single access restricted!
- Added unit test for the ReportSorter methods
- Changed the getWeekTitle method to properly extract week and year from the getReportWeek return value

Commit:
Added stream exception handling to Report class, added sorting implementation

- Added exception handling to Report class for when stream handling fails
- Added logging
- Implemented Report::sortSectionsBy - ReportSorter methods are chosen according to the given ReportSortType

Commit:
Moved sorting functions into Report class

- Removed ReportSorter Interface and moved its functionality into Report class
- Implemented Report class tests

Commit:
Changed variable types for private Report methods

Commit:
Refactored class Report to interface ReportSectionGenerator

- This was done  order to avoid non reusability issues
- New Interface presents one public function which takes a Progress stream and returns a ReportSection List.

Commit:
Added new PDF creation class

- Added ReportEncoderFop:
- This is an alternative document creation abstraction similar to the ReportGenerator classes
- In contrast to those, ReportEncoderFop takes String arguments for adding content and does no specific formatting, with the exception of printing the input of addSectionTitle in bold and larger font
- Added new xsl and xml templates for use with ReportFormatEncoderFop
- Added unit tests for ReportFormatEncoderFop

- Changed ReportSection class to only handle String content
- Updated ReportSectionGenerator class accordingly
-

Commit:
Switched Report composition to new style

- Composing a team report now uses the newly added classes
- Added function to Report Composer to support single user reports
- Added user id access validation method to Reports (to verifiy a user  only  get single user reports of themself)

Commit:
Added sub filtering to new report generation style

- Each section will now be sorted by (several) other fields in addition to the primary sorting field

Commit:
Added option to sort ReportSections by several fields

- In addition to the overall section order, the interface now provides the option to determine the sorting hierarchy further

Commit:
Changed report formatting

- Reduced font size of section titles
- Added additional new lines to improve readability of the report

Commit:
Added REST endpoint to accessing user reports

- Every user can now access user reports pertaining to their own Progress entries
- In the future the team lead access will be restricted to users which are in at least one of their teams

Commit:
Extended report generation by sorting parameter

- The report generation endpoints now take an additional parameter denoting in which way the report data should be sorted

Commit:
Updated the dart sources to new sorting scheme, added user report page

- Updated navigation to present a way to user reports
- Refactored report widget:
- Is now statelss
- Contains the new widgets for selecting the sorting of reports (ReportRadioSort) and displaying the data sources (ReportTableManager)
- Sets if ReportTableManager and ReportRadioSort are displaying a user or team report

- Changed page report to be a stateless widget

Commit:
Added new security filter for using the flutter debugger

- When using the flutter debugger chrome interface, we essentially use cross site access to the backend
- This means out cookies need to have the parameters for cross site security set accordingly
- This is handled by the new filter JSESSION_CookieFilter
- When added to the filter chain it will intercept the JSESSION cookie and change/add the appropriate parameters

- Due to the way the JSESSION cookie is added initially, there is a usability issue:
- After the first log in the user should reload the site and then log in again
- This is due to the fact, that the JSESSION cookie is initially set after our filter and only in subsequent response added before the filter
- This means the first cookie will be issued without the proper parameters and the browser will find the cookie security lacking and not use it
- After the first use however, the correct parameters will be set and everything should work fine

Commit:
Report mails containing no progress will no longer be sent

- Added check into ReportGeneratorService:
- If the ReportSection List returned from the report composer is empty, the report will not be finalized and no mail will be dispatched
- Added check into ReportSectionGenerator to avoid appending an empty section to the output

Commit:
Upgraded to spring boot version 2.5.4

- Updated tests to use jupiter because annotation @RunWith has been removed

Commit:
Fixed use of Logger in Jsession cookie filter

Commit:
Refined structure of report sections

- The body of a Report section is now a List of ReportSectionBody instances
- ReportSectionBody contains fields for the progress test, its title as well as the progress meta information

Commit:
Updated year in copyright header




+

Commit:
Beautified ReportSectionGenerator source file

Commit:
Switched up Jsession cookie flag setting

- During development: the cookie flag "SameSite" is now set by defining a TomcatContextCustomizer bean
- Previously a filter intercepted the cookie and changed its flags
- Added attribute cookie flag secure in application.attributes. This was previously done by the filter as well.

Commit:
Changed interface of ReportGenerator

- Switched from an interface definition to a class
- Instances are returned by build
- Functions sortBy and withSortFieldValues configure an instance
- The ReportSections are generated by calling create

botorabi commented on src/main/java/net/vrfun/tasktracker/report/ReportSectionGenerator.java:
```diff
@@ -0,0 +1,279 @@
+    List<ReportSection> create() {
+        return sortByFields();
+    }
+
```
No need for 'static', it is valid for the other methods, too.

botorabi commented on src/main/java/net/vrfun/tasktracker/report/Reports.java:
```diff
@@ -64,6 +65,25 @@ public boolean validateUserAccess(@NonNull final List<Long> teamIDs) {
+
+            Optional<UserDTO> userDTO = userRepository.getUserById(userId);
+            boolean returnValue = false;
+            if (userDTO.isPresent()) {
```
Use 'equals' instead of '==' for comparing Longs

botorabi commented on src/main/java/net/vrfun/tasktracker/report/docgen/ReportEncoderFop.java:
```diff
@@ -0,0 +1,200 @@
+        setHeader("");
+        setFooter("");
+    }
+
```
Make this method private

botorabi commented on src/main/java/net/vrfun/tasktracker/report/docgen/ReportDocumentCreator.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
```
Use a class instead of interface, use a static  build() method and make the other methods non-static.

botorabi commented on src/main/java/net/vrfun/tasktracker/security/JSESSION_CookieFilter.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Collection;
+import java.util.List;
+
+
```
No need for this class, drop it.

botorabi commented on src/main/java/net/vrfun/tasktracker/report/ReportSectionGenerator.java:
```diff
@@ -0,0 +1,279 @@
+    /**
+     * Sorts output
+     * @param sortByField Field by which the progresses will be sorted
+     */
```
private?

Commit:
Code cleanup, unit test fix

- Added annotation @NonNull as well as final signifier where applicable
- Updated code formatting
- Fixed unit test when sorting reports by user
- Removed unsused sources

botorabi committed:
Added support for docker image creation

botorabi committed:
Made the DB settings coming from env

Commit:
Updated access modifiers for class ReportSectionGenerator
